main.go

```diff
@@ -44,18 +52,93 @@ func main() {
 		os.Exit(1)
 	}
 
-	svc := &service.Service{
+	dimensionExtractedErrProducer, err := kafka.NewProducer(cfg.Brokers, cfg.EventReporterTopic, int(envMax))
+	if err != nil {
+		log.Error(err, nil)
+		os.Exit(1)
+	}
+
+	signals := make(chan os.Signal, 1)
+	signal.Notify(signals, syscall.SIGINT, syscall.SIGTERM)
+
+	eventLoopDone := make(chan bool)
+	apiErrors := make(chan error, 1)
+
+	api.CreateDimensionExtractorAPI(cfg.DimensionExtractorURL, cfg.BindAddr, apiErrors)
+
+	service := &service.Service{
 		EnvMax:                envMax,
-		BindAddr:              cfg.BindAddr,
-		Consumer:              syncConsumerGroup,
 		DatasetAPIURL:         cfg.DatasetAPIURL,
 		DatasetAPIAuthToken:   cfg.DatasetAPIAuthToken,
 		DimensionExtractorURL: cfg.DimensionExtractorURL,
 		MaxRetries:            cfg.MaxRetries,
 		Producer:              dimensionExtractedProducer,
 		S3:                    s3,
-		Shutdown:              cfg.GracefulShutdownTimeout,
+		HTTPClient:            rchttp.DefaultClient,
+	}
+
+	errorReporter, err := reporter.NewImportErrorReporter(dimensionExtractedErrProducer, log.Namespace)
+	if err != nil {
+		log.ErrorC("error while attempting to create error reporter client", err, nil)
+		os.Exit(1)
+	}
+
+	eventConsumer := event.Consumer{
+		KafkaConsumer: syncConsumerGroup,
+		EventService:  service,
+		ErrorReporter: errorReporter,
+	}
+
+	eventLoopContext, eventLoopCancel := context.WithCancel(context.Background())
+	eventConsumer.Start(eventLoopDone, eventLoopContext)
+
+	// block until a fatal error, signal or eventLoopDone - then proceed to shutdown
+	select {
+	case <-eventLoopDone:
+		log.Debug("quitting after done was closed", nil)
+	case signal := <-signals:
+		log.Debug("quitting after os signal received", log.Data{"signal": signal})
+	case consumerError := <-syncConsumerGroup.Errors():
+		log.Error(fmt.Errorf("aborting consumer"), log.Data{"message_received": consumerError})
+	case producerError := <-service.Producer.Errors():
+		log.Error(fmt.Errorf("aborting producer"), log.Data{"message_received": producerError})
+	case <-apiErrors:
+		log.Error(fmt.Errorf("server error forcing shutdown"), nil)
 	}
 
-	svc.Start()
+	// give the app `Timeout` seconds to close gracefully before killing it.
+	ctx, cancel := context.WithTimeout(context.Background(), cfg.GracefulShutdownTimeout)
+
+	go func() {
+		log.Debug("stopping kafka consumer listener", nil)
+		syncConsumerGroup.StopListeningToConsumer(ctx)
+		log.Debug("stopped kafka consumer listener", nil)
+		eventLoopCancel()
+		<-eventLoopDone
+		log.Debug("closing http server", nil)
+		if err := api.Close(ctx); err != nil {
+			log.ErrorC("failed to gracefully close http server", err, nil)
+		} else {
+			log.Debug("gracefully closed http server", nil)
+		}
+		log.Debug("closing kafka producer", nil)
+		service.Producer.Close(ctx)
+		log.Debug("closed kafka producer", nil)
```

Comment:
We should be gracefully closing the `dimensionExtractedErrProducer` between lines 126 and 127.

```
log.Debug("closing kafka dimension extracted error producer", nil)
dimensionExtractedErrProducer.Close(ctx)
log.Debug("closed kafka dimension extracted error producer", nil)